File: data/src/app/workspace.kt
package data.app

import androidx.room.ColumnInfo
import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Entity
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.PrimaryKey
import androidx.room.Query
import androidx.room.Update
import kotlinx.coroutines.flow.Flow
import kotlinx.serialization.Serializable

@Entity(
    tableName = "workspaces"
) @Serializable data class Workspace(
    @PrimaryKey val id: String,
    val name: String,
    @ColumnInfo(name = "owner_id") val ownerId: String,
    @ColumnInfo(name = "created_at", defaultValue = "CURRENT_TIMESTAMP") val createdAt: String = "",
    @ColumnInfo(name = "updated_at", defaultValue = "CURRENT_TIMESTAMP") val updatedAt: String = "",
    @ColumnInfo(defaultValue = "0") val deleted: Boolean = false
)

@Dao interface WorkspaceDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE) suspend fun upsert(item: Workspace)
    @Update suspend fun update(item: Workspace)
    @Delete suspend fun delete(item: Workspace)

    @Query("UPDATE workspaces SET deleted = 1, updated_at = CURRENT_TIMESTAMP WHERE id = :id")
    suspend fun softDelete(id: String)

    @Query("SELECT * FROM workspaces WHERE updated_at > :timestamp")
    suspend fun getChanges(timestamp: String): List<Workspace>

    @Query("SELECT * FROM workspaces WHERE id = :id")
    suspend fun getById(id: String): Workspace?

    @Query("SELECT * FROM workspaces")
    fun getAll(): Flow<List<Workspace>>
}

File: server/data/app/workspace.ts
import { sqliteTable, text, integer, index } from "drizzle-orm/sqlite-core"
import { eq, sql } from "drizzle-orm"
import { DrizzleD1Database } from "drizzle-orm/d1"
import { User } from "./user"

export const Workspace = sqliteTable(
  "workspaces",
  {
    created_at: text("created_at").default(sql`CURRENT_TIMESTAMP`),
    deleted: integer("deleted", { mode: "boolean" }).default(false),
    id: text("id").primaryKey(),
    name: text("name"),
    owner_id: text("owner_id").references(() => User.id, { onDelete: "restrict" }),
    updated_at: text("updated_at").default(sql`CURRENT_TIMESTAMP`),
  },
  (table) => ({
    idx0: index("workspaces_owner_id_idx").on(table.owner_id),
  }),
)

export class WorkspaceDao {
  constructor(private db: DrizzleD1Database<any>) {}

  async upsert(item: typeof Workspace.$inferInsert) {
    return this.db
      .insert(Workspace)
      .values(item)
      .onConflictDoUpdate({ target: Workspace.id, set: item })
      .run()
  }

  async update(item: Partial<typeof Workspace.$inferInsert> & { id: any }) {
    return this.db.update(Workspace).set(item).where(eq(Workspace.id, item.id!)).run()
  }

  async delete(item: { id: any }) {
    return this.db.delete(Workspace).where(eq(Workspace.id, item.id!)).run()
  }

  async getById(id: string) {
    return this.db.select().from(Workspace).where(eq(Workspace.id, id)).get()
  }

  async getAll() {
    return this.db.select().from(Workspace).all()
  }

  async getChanges(timestamp: string) {
    return this.db
      .select()
      .from(Workspace)
      .where(sql`${Workspace.updated_at} > ${timestamp}`)
      .all()
  }
}
